### PR TITLE
Bingads

### PR DESCRIPTION
**Pull request by 1abhishekpandey**
## Description of the change

> Provided support for BingAds in JS-sdk.

## Type of change
- [ ] New feature (non-breaking change that adds functionality)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

---
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/226)

### CODE REVIEW

File: integrations/BingAds/browser.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+  /*
+    Visit here(for details Parameter details): https://help.ads.microsoft.com/#apex/3/en/53056/2
+    Under: What data does UET collect once I install it on my website?
+    It conatins info about parameters ea,ec,gc,gv,el
+
+  */
+
+  track(rudderElement) {
+    const typeofcall = rudderElement.message.type;
```
Comment (ruchiramoitra) [inline integrations/BingAds/browser.js:49]: lets destructure this. 
const { type, properties } = rudderElement.message
Also, destructure the properties
const {category,currency,etc) = properties

File: integrations/BingAds/browser.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    const properties = rudderElement.message.properties;
+    var event = {
+      ea: typeofcall
+    };
+    if (properties.category) event.ec = properties.category;
```
Comment (ruchiramoitra) [inline integrations/BingAds/browser.js:54]: lets use like
event.ec = category ? category : null

File: integrations/BingAds/browser.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    if (properties.currency) event.gc = properties.currency;
+    if (properties.value) event.gv = properties.value;
+    if (properties.label) event.el = properties.label;
```
Comment (ruchiramoitra) [inline integrations/BingAds/browser.js:57]: don't we need ea? (event action)

File: integrations/BingAds/browser.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      }, n = d.createElement(t), n.src = r, n.async = 1, n.onload = n.onreadystatechange = function() {
+        var s = this.readyState;
+        s && s !== "loaded" && s !== "complete" || (f(), n.onload = n.onreadystatechange = null)
+      }, i = d.getElementsByTagName(t)[0], i.parentNode.insertBefore(n, i)
+    })(window, document, "script", "https://bat.bing.com/bat.js", "uetq");
+  }
+
+  init() {
+    logger.debug("===in init BingAds===");
+    console.log(window);
```
Comment (ruchiramoitra) [inline integrations/BingAds/browser.js:27]: remove console

File: integrations/BingAds/browser.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
+import logger from "../../utils/logUtil";
+import ScriptLoader from "../ScriptLoader";
+
+class BingAds {
+  constructor(config) {
+    this.apikey = config.apikey;
+  }
+
+  loadBingadsScript() {
+    var apikey = this.apikey;
```
Comment (ruchiramoitra) [inline integrations/BingAds/browser.js:10]: lets follow es6. dont use var. here use destructuring. 
const.

File: integrations/BingAds/browser.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
+import logger from "../../utils/logUtil";
+import ScriptLoader from "../ScriptLoader";
```
Comment (ruchiramoitra) [inline integrations/BingAds/browser.js:2]: remove this import

File: integrations/BingAds/browser.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    })(window, document, "script", "https://bat.bing.com/bat.js", "uetq");
+  }
+
+  init() {
+    logger.debug("===in init BingAds===");
+    console.log(window);
+    this.loadBingadsScript();
+  }
+
+  isLoaded() {
```
Comment (ruchiramoitra) [inline integrations/BingAds/browser.js:31]: isLoaded = () => use this type of syntax for all the function. read about this too.

File: integrations/BingAds/browser.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+      ea: typeofcall
+    };
+    if (properties.category) event.ec = properties.category;
+    if (properties.currency) event.gc = properties.currency;
+    if (properties.value) event.gv = properties.value;
+    if (properties.label) event.el = properties.label;
+    window.uetq.push(event);
+  }
+
+  page(rudderElement) {
```
Comment (ruchiramoitra) [inline integrations/BingAds/browser.js:61]: no need to pass rudderelement as it is not used in the function

File: integrations/BingAds/browser.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    Visit here(for details Parameter details): https://help.ads.microsoft.com/#apex/3/en/53056/2
+    Under: What data does UET collect once I install it on my website?
+    It conatins info about parameters ea,ec,gc,gv,el
+
+  */
+
+  track(rudderElement) {
+    const typeofcall = rudderElement.message.type;
+    const properties = rudderElement.message.properties;
+    var event = {
```
Comment (ruchiramoitra) [inline integrations/BingAds/browser.js:51]: dont use var

**Review by ruchiramoitra — APPROVED** [review]

File: integrations/BingAds/browser.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    It conatins info about parameters ea,ec,gc,gv,el
+  */
+
+  track = (rudderElement) => {
+    const { type, properties } = rudderElement.message;
+    const { category, currency, value, label } = properties;
+    let event = {
+      ea: type
+    };
+    if (category) event.ec = category;
```
Comment (arnab-p) [inline integrations/BingAds/browser.js:50]:
```suggestion
    if (category) { 
      event.ec = category;
    }
```

For better readability. Also, apply the same to others.

**Review by arnab-p — CHANGES_REQUESTED** [review]

**Review by arnab-p — APPROVED** [review]

File: integrations/BingAds/browser.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    let event = {
+      ea: type
+    };
+    if (category) {
+      event.ec = category;
+    }
+    if (currency) {
+      event.gc = currency;
+    }
+    if (value) {
```
Comment (sayan-mitra) [inline integrations/BingAds/browser.js:56]: this seems like revenue value.
Do we support other keys where revenue can come from?

Comment (1abhishekpandey) [reply integrations/BingAds/browser.js:56]: Yes, this is revenue value, it is used to set BingAds **gv** (Variable revenue currency) parameter.
For now, we support only one key.
More detail: Revenue comes from properties fields: e.g.,
`rudderanalytics.track('track conversion', {
                category: 'MyCategory',
                value: 125,
                currency: 'INR',
                order_id: 'order_1'
            });`
What changes will I need to do, please suggest.

Comment (sayan-mitra) [reply integrations/BingAds/browser.js:56]: I think revenue and total can also be considered as valid keys.
@ruchiramoitra plz confirm

Comment (ruchiramoitra) [reply integrations/BingAds/browser.js:56]: yes we should consider them too... value, revenue, total

File: integrations/BingAds/browser.js [hunk 12]
```diff
@@ -56,6 +56,13 @@ class BingAds {
     if (value) {
       event.gv = value;
     }
+    else if(revenue) {
```
Comment (sayan-mitra) [inline integrations/BingAds/browser.js:59]: let's use if's only to keep parity with others.
Also update the doc accordingly, the highest preference is total > revenue > value

Comment (1abhishekpandey) [reply integrations/BingAds/browser.js]: Please check it once more. If any other changes is needed or not.

**Review by sayan-mitra — APPROVED** [review]

**Pull request merged**